influxdb3_py_api/src/system_py.rs

```diff
@@ -899,6 +939,310 @@ fn process_response_part(
     Ok((response_str, headers))
 }
 
+// Cache entry with optional expiration
+struct CacheEntry {
+    value: Py<PyAny>,         // Python object reference with proper reference counting
+    expires_at: Option<Time>, // Expiration time if any
+}
+
+impl CacheEntry {
+    fn new(value: Py<PyAny>, ttl: Option<f64>, time_provider: Arc<dyn TimeProvider>) -> Self {
+        let expires_at = ttl.map(|seconds| time_provider.now() + Duration::from_secs_f64(seconds));
+        Self { value, expires_at }
+    }
+}
+
+// Cache store that manages namespaced caches
+struct CacheStore {
+    // Map of namespace -> (key -> cache entry)
+    namespaces: HashMap<CacheId, ExpiringCache>,
+    time_provider: Arc<dyn TimeProvider>,
+    last_cleanup: Time,
+    cleanup_interval: Duration,
+}
+
+struct ExpiringCache {
+    entries: HashMap<String, CacheEntry>,
+    expirations: BTreeMap<Time, HashSet<String>>,
+    default_ttl: Option<Duration>,
+    time_provider: Arc<dyn TimeProvider>,
+}
+
+impl ExpiringCache {
+    fn new(time_provider: Arc<dyn TimeProvider>, default_ttl: Option<Duration>) -> Self {
+        Self {
+            entries: HashMap::default(),
+            expirations: BTreeMap::default(),
+            default_ttl,
+            time_provider,
+        }
+    }
+
+    fn insert(&mut self, key: String, mut entry: CacheEntry) {
+        // this is for test call caches, which should expire.
+        if self.default_ttl.is_some() && entry.expires_at.is_none() {
+            entry.expires_at = Some(self.time_provider.now() + self.default_ttl.unwrap());
+        }
+        // if key has a ttl, record its expiration.
+        let expiration = entry.expires_at;
+        // if this was an overwrite, remove from expirations
+        if let Some(old_entry) = self.entries.insert(key.clone(), entry) {
+            if let Some(old_expiration) = old_entry.expires_at {
+                self.expirations
+                    .get_mut(&old_expiration)
+                    .unwrap()
+                    .remove(&key);
+            }
+        }
+        if let Some(expiration) = expiration {
+            self.expirations.entry(expiration).or_default().insert(key);
+        }
+    }
+
+    fn get(&mut self, key: &str, py: Python<'_>) -> Option<Py<PyAny>> {
+        let entry = self.entries.get(key)?;
+        if let Some(expiration) = entry.expires_at {
+            if expiration <= self.time_provider.now() {
+                self.remove(key);
+                return None;
+            }
+        }
+        Some(entry.value.clone_ref(py))
+    }
+
+    fn remove(&mut self, key: &str) -> bool {
+        if let Some(entry) = self.entries.remove(key) {
+            if let Some(expiration) = entry.expires_at {
+                self.expirations.get_mut(&expiration).unwrap().remove(key);
+            }
+            return true;
+        }
+        false
+    }
+
+    fn cleanup(&mut self) {
```

Comment:
this function seems like it's going to be problematic if we get into a situation with many keys in the cache. Might not be a problem, but something to look out for. Figuring out some way to look for expirations without holding a write lock might be good for later.